### PR TITLE
fix: `multi_scalar_multiplication`

### PR DESCRIPTION
**Pull request by ratankaliani**
Handle the complete addition for `secp256k1_add` using a new `WeierstrassAffinePoint` trait that's implemented for secp256k1, bls12-381 and bn254. Easy to extend to `secpr1`.

- Modify `multi_scalar_multiplication` and `mul_assign` to use `complete_add_assign`, which handles all of the potential edge cases. 
- Add generic test for `secp256k1_add`, `bls12381_add` and `bn254_add` that covers the complete addition cases.

### CODE REVIEW

File: tests/common/src/weierstrass_add.rs
Comment (ratankaliani) [inline tests/common/src/weierstrass_add.rs:1]: Add a test verifying that P + (order of curve - 1) * P = point at infinity

File: crates/zkvm/lib/src/utils.rs [hunk 1]
```diff
@@ -130,3 +139,57 @@ pub fn bytes_to_words_le(bytes: &[u8]) -> Vec<u32> {
         .map(|chunk| u32::from_le_bytes(chunk.try_into().unwrap()))
         .collect::<Vec<_>>()
 }
+
+/// A trait for affine points on Weierstrass curves.
+pub trait WeierstrassAffinePoint<const N: usize>: AffinePoint<N> {
+    /// The infinity point is set to (0, 0).
+    fn infinity() -> Self {
```
Comment (tamirhemo) [inline crates/zkvm/lib/src/utils.rs:146]: This seems kind of bad? it would be good if an affine point is valid.

Comment (ratankaliani) [reply crates/zkvm/lib/src/utils.rs]: Now represented as a separate type.

File: crates/zkvm/lib/src/secp256k1.rs [hunk 2]
```diff
@@ -1,12 +1,25 @@
...
+};
 
 /// The number of limbs in [Secp256k1AffinePoint].
 pub const N: usize = 16;
 
 /// An affine point on the Secp256k1 curve.
 #[derive(Copy, Clone)]
 #[repr(align(4))]
-pub struct Secp256k1AffinePoint(pub [u32; N]);
+pub struct Secp256k1AffinePoint(pub WeierstrassPoint<N>);
```
Comment (tamirhemo) [inline crates/zkvm/lib/src/secp256k1.rs:12]: nit: rename to Secp256k1Point?

**Review by tamirhemo — APPROVED** [review]
